## Built in Functions:
The functions which are coming along with Python software automatically, are called built in functions or pre defined functions

    Eg:
    id()
    type()
    input()
    eval()
    etc..

## a. input()
This function always reads the data from the keyboard in the form of String Format. We
have to convert that string type to our required type by using the corresponding type
casting methods.

    The input() method reads the user's input and returns it as a string
    

In [1]:
# example
name = input('Enter your name: ')
print(name)

# It will always print str type only for any input type

Enter your name: Nikhil Yadav
Nikhil Yadav


In [2]:
Age  = input('Please enter the age: ')
print(Age)
print(type(Age))

Please enter the age: 20
20
<class 'str'>


In [3]:
print(name,Age, sep=' --- ')

Nikhil Yadav --- 20


In [4]:
# type casting
age = int(Age)
print(type(age))

<class 'int'>


#### Q. Write a program to read 2 numbers from the keyboard and print sum.

In [5]:
n1 = int(input('Enter the first number: '))
n2 = int(input('Enter the second number: '))
print('Sum of two numbers: ', n1+n2)

Enter the first number: 12
Enter the second number: 24
Sum of two numbers:  36


#### Q. Write a program to read Student data from the keyboard and print that data.

In [6]:
name = input("Enter the student's name: ")
roll_number = int(input('Roll number: '))
address = input('Enter the address: ')
course = input('Enter the course: ')
print('---------------------------------')

print("Please Confirm Information") 
print('Student Name',name)
print('Student Roll no.',roll_number)
print('Student Address',address)
print('Course Name',course)

Enter the student's name: Nikhil Yadav
Roll number: 34
Enter the address: Bulandshahr
Enter the course: MCA
---------------------------------
Please Confirm Information
Student Name Nikhil Yadav
Student Roll no. 34
Student Address Bulandshahr
Course Name MCA


## b. split()

The split() method splits a string into a list. You can specify the separator, default separator is any whitespace.

#### How to read multiple values from the keyboard in a single line:

In [7]:
a,b= [int(x) for x in input("Enter 2 numbers: ").split()]
print("Product is :", a*b) 

Enter 2 numbers: 8 5
Product is : 40


Note: split() function can take space as seperator by default .But we can pass anything as seperator.


#### Q. Write a program to read 3 float numbers from the keyboard with , seperator and print their sum.


In [8]:
a,b,c= [float(x) for x in input("Enter 3 float numbers :").split(',')]
print("The Sum is :", a+b+c) 

Enter 3 float numbers :12.4,5.4,6.5
The Sum is : 24.3


#### Q: "how are you dear" print this string in the given five ways. 

In [9]:
x1 = "how are you dear"
for i  in x1.split()[::-1]:
    print(i,end=" ")

dear you are how 

In [10]:
x2 = "how are you dear"
for i in x2.split():
    print(i[::-1],end=" ")

woh era uoy raed 

In [11]:
x3 = "how are you dear"
for i in x3.split():
    print(i[:-1]+i[-1].upper(),end=" ")

hoW arE yoU deaR 

In [12]:
x4 = "how are you dear"
for i in x4.split():
    print(i[:-2]+i[-2].upper()+i[-1],end=" ")

hOw aRe yOu deAr 

In [13]:
x5 = "how are you dear"
for i in x5.split():
    print(i[0].upper()+i[1:-1]+i[-1].upper(),end=" ")

HoW ArE YoU DeaR 

#### Q: Sum of all marks in given string

In [14]:
s = "English 70 Physics 65 Chemistry 77 Mathematics 62 "
x = 0
for i in s.split():
    if i.isdigit():
        x  = x + int(i)
x

274

## c. eval():

Eval Function take a String and evaluate the Result.


In [15]:
# example
x = eval('10+20+30')
print(x)

60


In [16]:
x = eval(input('Enter Expression'))
print(x)

Enter Expression2*5/6**2+9
9.277777777777779


eval() can evaluate the Input to list, tuple, set, etc based the provided Input.
#### Eg: Write a Program to accept list from the keynboard on the display

In [17]:
l = eval(input('Enter list: '))
print (type(l))
print(l)

Enter list: 12,34,5,3,2,1,3,5
<class 'tuple'>
(12, 34, 5, 3, 2, 1, 3, 5)


## Output statements:

We can use print() function to display output.


#### Form-1: print() without any argument
Just it prints new line character

#### Form-2: print('string')

In [18]:
# example
print("Hello World")

# We can use escape characters also
print("Hello \n World")
print("Hello\tWorld")

# We can use repetetion operator (*) in the string
print(3*"Hello")
print("Hello"*2)

# We can use + operator also
print("Hello"+"World") 

Hello World
Hello 
 World
Hello	World
HelloHelloHello
HelloHello
HelloWorld


#### Note:

    If both arguments are String type then + operator acts as concatenation operator.
    If one argument is string type and second is any other type like int then we will get Error
    If both arguments are number type then + operator acts as arithmetic addition operator
#### Note:

In [19]:
print("Hello"+"World")
print("Hello","World")

HelloWorld
Hello World


#### Form-3: print() with variable number of arguments:


In [20]:
# example
a,b,c=10,20,30
print("The Values are :",a,b,c) 

The Values are : 10 20 30


#### By default output values are seperated by space. If we want we can specify seperator by using "sep" attribute:

In [21]:
# example
a,b,c=10,20,30
print(a,b,c,sep=',')
print(a,b,c,sep=':') 

10,20,30
10:20:30


#### Form-4: print() with end attribute:

In [22]:
print("Hello")
print("Nikhil")
print("Yadav") 

Hello
Nikhil
Yadav


If we want output in the same line with space

In [23]:
# example
print("Hello",end=' ')
print("Nikhil",end=' ')
print("Yadav") 

Hello Nikhil Yadav


Note: The default value for end attribute is \n,which is nothing but new line character.

#### Form-5: print(object) statement:
We can pass any object (like list,tuple,set etc)as argument to the print() statement. 

In [24]:
l=[10,20,30,40]
t=(10,20,30,40)
print(l)
print(t) 

[10, 20, 30, 40]
(10, 20, 30, 40)


#### Form-6: print(String,variable list):
We can use print() statement with String and any number of arguments.

In [25]:
# example
s="Nikhil"
a=20
s1="Machine Learning"
s2="Python"
print("Hello",s,"Your Age is",a)
print("You are teaching",s1,"and",s2)

Hello Nikhil Your Age is 20
You are teaching Machine Learning and Python


#### Form-7: print(formatted string):
    %i====>int
    %d====>int
    %f=====>float
    %s======>String type


In [26]:
# Syntax:

#print("formatted string" %(variable list))

# Eg 1:
a=10
b=20
c=30
print("a value is %i" %a)
print("b value is %d and c value is %d" %(b,c))


a value is 10
b value is 20 and c value is 30


In [27]:
# Eg 2:
s="Nikhil"
list=[10,20,30,40]
print("Hello %s ...The List of Items are %s" %(s,list))

Hello Nikhil ...The List of Items are [10, 20, 30, 40]


#### Form-8: print() with replacement operator { }

In [28]:
name="Nikhil"
salary=10000
gf="Prakriti"

print(f"Hello {name} your salary is {salary} and Your Friend {gf} is waiting")
print("Hello {} your salary is {} and Your Friend {} is waiting".format(name,salary,gf))
print("Hello {0} your salary is {1} and Your Friend {2} is waiting".format(name,salary,gf))
print("Hello {x} your salary is {y} and Your Friend {z} is waiting".format(x=name,y=salary,z=gf))

Hello Nikhil your salary is 10000 and Your Friend Prakriti is waiting
Hello Nikhil your salary is 10000 and Your Friend Prakriti is waiting
Hello Nikhil your salary is 10000 and Your Friend Prakriti is waiting
Hello Nikhil your salary is 10000 and Your Friend Prakriti is waiting


__________________________________________________________________________________________________

![](images\flow_control.png)

### I. Conditional Statements

### 1) if
    if condition : statement
     or
    if condition :
     statement-1
     statement-2
     statement-3
    If condition is true then statements will be executed.


In [29]:
# example
name=input("Enter Name: ")
if name=="Nikhil" :
    print("Hello Nikhil Good Morning")
print("How are you!!!") 

Enter Name: Nikhil Yadav
How are you!!!


### 2) if - else:

    if condition :
     Action-1
    else :
     Action-2
If condition is true then Action-1 will be executed otherwise Action-2 will be executed.

In [30]:
# example
name=input("Enter Name: ")
if name=="nikhil" :
    print("Hello Nikhil Good Morning")
else:
    print("Hello Good Moring")
    print("How are you!!!")

Enter Name: Prakriti
Hello Good Moring
How are you!!!


___elif in python means [else if].___



### 3) if - elif - else:
#### Syntax:
    if (condition 1):  ----> if condition 1 is true.
        print("yes")

    elif (condition 2):  ----> if condition 2 is true.
    print("no")
    
    else:
        print("may be")
    
    OR

    if condition1:
     Action-1
    elif condition2:
     Action-2
    elif condition3:
     Action-3
    elif condition4:
     Action-4
     ...
    else:
     Default Action

Based condition the corresponding action will be executed.

In [31]:
# example
subject = input("Enter Your Favourite Subject: ")

if subject=="Python" :
    print("Python is a popular programming language.")
elif subject=="ML":
    print("Machine learning (ML) is a field of inquiry devoted to understanding and building methods")
elif subject=="AI":
    print('''Artificial intelligence (AI) is intelligence demonstrated by machines, 
      as opposed to the natural intelligence displayed by animals including humans.''')
else :
    print("Other Subject are not recommended") 

Enter Your Favourite Subject: ML
Machine learning (ML) is a field of inquiry devoted to understanding and building methods


#### Note:

1. else part is always optional
 Hence the following are various possible syntaxes.
 1. if
 2. if - else
 3. if-elif-else
 4.if-elif
 
2. There is no switch statement in Python

#### Q. Write a program to find biggest of given 2 numbers from the commad prompt?

In [32]:
n1=int(input("Enter First Number:"))
n2=int(input("Enter Second Number:"))
if n1>n2:
    print("Biggest Number is:",n1)
else :
    print("Biggest Number is:",n2) 

Enter First Number:12
Enter Second Number:8
Biggest Number is: 12


#### Q. Write a program to find biggest of given 3 numbers from the commad prompt?

In [33]:
a = int(input("Enter the number "))
b = int(input("Enter the number "))
c = int(input("Enter the number "))
d = int(input("Enter the number "))

if a>b and a>c and a>d:
    print("a is greater",a)
elif b>a and b>c and b>d:
    print("b is greater",b)
elif  c>a and c>b and c>d:
    print("c is greater",c)
else:
    print("d is greater",d)

Enter the number 2
Enter the number 3
Enter the number 4
Enter the number 5
d is greater 5


Q. Write a program to find smallest of given 2 numbers?

Q. Write a program to find smallest of given 3 numbers?

Q. Write a program to check whether the given number is even or odd?

Q. Write a program to check whether the given number is in between 1 and 100?

## II. Iterative Statements
If we want to execute a group of statements multiple times then we should go for Iterative statements.

Python supports 2 types of iterative statements.
1. for loop
2. while loop


### 1) for loop:
If we want to execute some action for every element present in some sequence(it may be string or collection)then we should go for for loop.

    THe syntax of a loop looks like this.
        for x in sequence :
            body
Where sequence can be string or any collection.
Body will be executed for every element present in the sequence.

#### Eg 1: To print characters present in the given string

In [34]:
s="Prakriti"
for x in s :
    print(x)

P
r
a
k
r
i
t
i


#### Eg 2: To print characters present in string index wise:

In [35]:
name = "nikhil"
for i in range(len(name)):
    print(i,"--->",name[i])

0 ---> n
1 ---> i
2 ---> k
3 ---> h
4 ---> i
5 ---> l


### 2) while loop:
If we want to execute a group of statements iteratively until some condition false, then we should go for while loop.

The syntax of a while loop looks like this:

    while condition:     -------> the block keeps executing until the conndition is true.
            #body of the loop
In while loop, the condition is checked first. If it evolutes to true, the body of the loop is executed, otherwise not!

If the loop is entered, the process of [condition check and execution] is continued until the condition become false.
    
#### Eg 1: To print numbers from 1 to 10 by using while loop

In [36]:
x=1
while x <=10:
    print(x)
    x=x+1

1
2
3
4
5
6
7
8
9
10


#### Eg 2: To display the sum of first n numbers


In [37]:
n=int(input("Enter number:"))
sum=0
i=1
while i<=n:
    sum=sum+i
    i=i+1
print("The sum of first",n,"numbers is :",sum)

Enter number:14
The sum of first 14 numbers is : 105


#### Eg 3: write a program to prompt user to enter some name until entering Prakriti: 

In [38]:
name=""
while name!="Prakriti":
    name=input("Enter Name:")
print("Thanks for confirmation") 

Enter Name:Nikhil
Enter Name:Akhil
Enter Name:Prakriti
Thanks for confirmation


### Nested Loops:
Sometimes we can take a loop inside another loop,which are also known as nested loops.

In [39]:
# Eg:
for i in range(4):
    for j in range(4):
        print("i=",i," j=",j) 

i= 0  j= 0
i= 0  j= 1
i= 0  j= 2
i= 0  j= 3
i= 1  j= 0
i= 1  j= 1
i= 1  j= 2
i= 1  j= 3
i= 2  j= 0
i= 2  j= 1
i= 2  j= 2
i= 2  j= 3
i= 3  j= 0
i= 3  j= 1
i= 3  j= 2
i= 3  j= 3


#### Q. Write a program to dispaly *'s in Right angled triangled form

In [40]:
n = int(input("Enter number of rows:"))
for i in range(1,n+1):
    for j in range(1,i+1):
        print("*",end=" ")
    print() 

Enter number of rows:6
* 
* * 
* * * 
* * * * 
* * * * * 
* * * * * * 


In [41]:
# Alternative way:
n = int(input("Enter number of rows:"))
for i in range(1,n+1):
    print("* " * i)

Enter number of rows:6
* 
* * 
* * * 
* * * * 
* * * * * 
* * * * * * 


#### Q. Write a program to display *'s in pyramid style(also known as equivalent triangle)

In [42]:
n = int(input("Enter number of rows:"))
for i in range(1,n+1):
    print(" " * (n-i),end="")
    print("* "*i) 

Enter number of rows:6
     * 
    * * 
   * * * 
  * * * * 
 * * * * * 
* * * * * * 


## III. Transfer Statements

### 1) Break:
'break' is used to come out of the loop when encountered. It instructs the program to- Exit the loop now.

In [43]:
# Eg:
for i in range(10):
    if i==7:
        print("Mission Successful: ")
        break
    print(i) 

0
1
2
3
4
5
6
Mission Successful: 


In [44]:
for i in range(0,80):
    print(i)  # -----> this will print 0,1,2,3,4,5
    if i == 5:
        break

0
1
2
3
4
5


## 2) Continue:
'continue' is used to stop the current iteration of the loop and continue with the next one. It instructs the program to "skin this iteration"

In [45]:
for i in range(4):
    if i == 2:  # ----> if i is 2, the tertion is skipped
        continue
    print(i)

0
1
3


In [46]:
items = [10,800,400,500,30]
for item in items:
    if item>=500:
        print("We cannot process this item :",item)
        continue
    print(item) 

10
We cannot process this item : 800
400
We cannot process this item : 500
30


In [47]:
numbers=[10,20,0,5,0,30]
for n in numbers:
    if n==0:
        print("We can't divide with zero... So, just skipping")
        continue
    print("100/{} = {}".format(n,100/n)) 

100/10 = 10.0
100/20 = 5.0
We can't divide with zero... So, just skipping
100/5 = 20.0
We can't divide with zero... So, just skipping
100/30 = 3.3333333333333335


### loops with else block:
Inside loop execution,if break statement not executed ,then only else part will be
executed.

else means loop without break

In [48]:
cart=[10,20,30,40,50]
for item in cart:
    if item>=500:
        print("We cannot process this order")
        break
        print(item)
    else:
        print("Congrats ...all items processed successfully")

Congrats ...all items processed successfully
Congrats ...all items processed successfully
Congrats ...all items processed successfully
Congrats ...all items processed successfully
Congrats ...all items processed successfully


#### Q. What is the difference between for loop and while loop in Python?
 
    We can use loops to repeat code execution
    Repeat code for every item in sequence ==>for loop
    Repeat code as long as condition is true ==>while loop
#### Q. How to exit from the loop?
    by using break statement
#### Q. How to skip some iterations inside loop?
    by using continue statement.
#### Q. When else part will be executed wrt loops?
    If loop executed without break


## 3) pass statement:
pass is a keyword in Python.In our programming syntactically if block is required which won't do anything then we can define that empty block with pass keyword.
        
    pass
     |- It is an empty statement
     |- It is null statement
     |- It won't do anything

    Eg:
    if True:
    SyntaxError: unexpected EOF while parsing
    if True: pass
    ==>valid
    def m1():
    SyntaxError: unexpected EOF while parsing

## Use case of pass:
Sometimes in the parent class we have to declare a function with empty body and child class responsible to provide proper implementation. Such type of empty body we can define by using pass keyword. (It is something like abstract method in java)

In [49]:
#Eg:
for i in range(100):
    if i%9==0:
        print(i)
    else:
        pass 

0
9
18
27
36
45
54
63
72
81
90
99


## Assert Statement
The assert statemant is useful to check if a particular condition is fulfilled or not.
    
    The syntax is as follows:
        
        assert expression, message
        
In the above syntax, the 'message' is not compulsory. Let's take an example. If we want to assure that the user should enter only a number greater than 0, we can use assert statement as: 

In [50]:
n = int(input("Enter the number: ")) 
assert n>0, "Wrong input entered"

Enter the number: 3


In [51]:
n = int(input("Enter the number: ")) 
assert n>0, "Wrong input entered"

Enter the number: -1


AssertionError: Wrong input entered

The 'AssertionError' shown in the above output is called an expection. An expection is an error that occurs during runtime. To avoid such type of expections, we can handel them using 'try ... expect' statement. After 'try', we use 'assert' and after 'except', we write the expection name to be handel. In the expect block, we write the statement that are executed in case of exception. 

In [52]:
# To handel AssertionError raised by assert
n = int(input("Enter the number: "))
try:
    assert n>0       # exception may occur here
    print("You Entered: ", n)
except AssertionError:
    print("Wrong Input Entered")      # this is executed in case of exception

Enter the number: -3
Wrong Input Entered


## del statement:
    del is a keyword in Python.
After using a variable, it is highly recommended to delete that variable if it is no longer
required,so that the corresponding object is eligible for Garbage Collection.

We can delete variable by using del keyword.

In [53]:
#Eg:
x=10
print(x)
del x

10


After deleting a variable we cannot access that variable otherwise we will get NameError.

    Eg:
       x=10
       del x
       print(x)
NameError: name 'x' is not defined.

___Note:
We can delete variables which are pointing to immutable objects.But we cannot delete
the elements present inside immutable object.___
    
    Eg:
        s="nikhil"
        print(s)
        del s==>valid
        del s[0] ==>TypeError: 'str' object doesn't support item deletion

### Difference between del and None:
In the case del, the variable will be removed and we cannot access that variable(unbind operation)

       s="nikhil"
       del s
       print(s) ==>NameError: name 's' is not defined.

But in the case of None assignment the variable won't be removed but the corresponding object is eligible for Garbage Collection(re bind operation). Hence after assigning with None value,we can access that variable.

    s="nikhil"
    s=None
    print(s) # None 

## FUNCTIONS

If a group of statements is repeatedly required then it is not recommended to write these statements everytime seperately.We have to define these statements as a single unit and we can call that unit any number of times based on our requirement without rewriting.

This unit is nothing but function.
The main advantage of functions is code Reusability.

___Note: In other languages functions are known as methods, procedures, subroutines etc.. Python supports 2 types of functions___
1. Built in Functions
2. User Defined Functions


### 1. Built in Functions:
The functions which are coming along with Python software automatically, are called built in functions or pre defined functions

    Eg:
    id()
    type()
    input()
    eval()
    etc..
    
### 2. User Defined Functions:
The functions which are developed by programmer explicitly according to business requirements, are called user defined functions.

Syntax to create user defined functions:
    
    def function_name(parameters) :

    """ doc string"""
     ----
     -----
     return value

___Note: While creating functions we can use 2 keywords___

1. def (mandatory)
2. return (optional)


In [56]:
# Eg 1: Write a function to print Hello
# test.py:
def wish():
    print("Hello Good Morning")

wish()
wish()
wish() 

Hello Good Morning
Hello Good Morning
Hello Good Morning


### Parameters :

Parameters are inputs to the function. If a function contains parameters,then at the time of calling,compulsory we should provide values otherwise,otherwise we will get error.

#### Eg: Write a function to take name of the student as input and print wish message by name.

In [57]:
def wish(name):
    print("Hello",name,"Good Morning")
wish("Nikhil")
wish("Yadav")

Hello Nikhil Good Morning
Hello Yadav Good Morning


#### Eg: Write a function to take number as input and print its square value

In [58]:
def squareIt(number):
    print("The Square of",number,"is", number*number)
squareIt(4)
squareIt(5) 

The Square of 4 is 16
The Square of 5 is 25


### Return Statement:
Function can take input values as parameters and executes business logic, and returns output to the caller with return statement.

#### Q. Write a function to accept 2 numbers as input and return sum.

In [59]:
def add(x,y):
    return x+y
result=add(10,20)
print("The sum is",result)
print("The sum is",add(100,200)) 

The sum is 30
The sum is 300


#### If we are not writing return statement then default return value is None

In [60]:
def f1():
    print("Hello")
f1()
print(f1()) 

Hello
Hello
None


#### Q. Write a function to check whether the given number is even or odd?

In [61]:
def even_odd(num):
    if num%2==0:
        print(num,"is Even Number")
    else:
        print(num,"is Odd Number")
even_odd(10)
even_odd(15) 

10 is Even Number
15 is Odd Number


#### Q. Write a function to find factorial of given number?

In [62]:
def fact(num):
    result=1
    while num>=1:
        result=result*num
        num=num-1
    return result
for i in range(1,5):
    print("The Factorial of",i,"is :",fact(i)) 

The Factorial of 1 is : 1
The Factorial of 2 is : 2
The Factorial of 3 is : 6
The Factorial of 4 is : 24


### Returning multiple values from a function:

In other languages like C,C++ and Java, function can return atmost one value. But in Python, a function can return any number of values.

In [63]:
# Eg 1:
def sum_sub(a,b):
    sum=a+b
    sub=a-b
    return sum,sub
x,y=sum_sub(100,50)
print("The Sum is :",x)
print("The Subtraction is :",y) 

The Sum is : 150
The Subtraction is : 50


In [64]:
# Eg 2:
def calc(a,b):
    sum=a+b
    sub=a-b
    mul=a*b
    div=a/b
    return sum,sub,mul,div
t=calc(100,50)
print("The Results are") 
for i in t:
    print(i)

The Results are
150
50
5000
2.0


### Types of arguments

    def f1(a,b):
     ------
     ------
     ------
    f1(10,20)
    
    a,b are formal arguments where as 10,20 are actual arguments
    
There are 4 types are actual arguments are allowed in Python.
1. positional arguments
2. keyword arguments
3. default arguments
4. Variable length arguments

### 1. positional arguments:
These are the arguments passed to function in correct positional order.

    def sub(a,b):
     print(a-b)
    sub(100,200)
    sub(200,100)
The number of arguments and position of arguments must be matched. If we change the order then result may be changed.

If we change the number of arguments then we will get error.

### 2. keyword arguments:
We can pass argument values by keyword i.e by parameter name.

In [65]:
def wish(name,msg):
    print("Hello",name,msg)
wish(name="Nikhil",msg="Good Morning")
wish(msg="Good Morning",name="Nikhil") 

Hello Nikhil Good Morning
Hello Nikhil Good Morning


___Here the order of arguments is not important but number of arguments must be matched___

#### Note: 
We can use both positional and keyword arguments simultaneously. But first we have totake positional arguments and then keyword arguments,otherwise we will get syntaxerror.

    def wish(name,msg):
        print("Hello",name,msg)
    wish("Nikhil","GoodMorning") ==>valid
    wish("Nikhil",msg="GoodMorning") ==>valid
    wish(name="Nikhil","GoodMorning") ==>invalid
    SyntaxError: positional argument follows keyword argument

### 3. Default Arguments:
Sometimes we can provide default values for our positional arguments.

In [66]:
# Eg:
def wish(name="Prakriti"):
    print("Hello",name,"Good Morning")
wish("Nikhil")
wish() 

Hello Nikhil Good Morning
Hello Prakriti Good Morning


___If we are not passing any name then only default value will be considered.___
#### Note:
After default arguments we should not take non default arguments
    
    def wish(name="Prakriti",msg="Good Morning"): ===>Valid
    def wish(name,msg="Good Morning"): ===>Valid
    def wish(name="Prakriti",msg): ==>Invalid
    SyntaxError: non-default argument follows default argument

### 4. Variable length arguments:
Sometimes we can pass variable number of arguments to our function,such type of arguments are called variable length arguments.

We can declare a variable length argument with * symbol as follows
    
    def f1(*n):
We can call this function by passing any number of arguments including zero number.
Internally all these values represented in the form of tuple.

In [67]:
# Eg:
def sum(*n):
    total=0
    for n1 in n:
        total=total+n1
    print("The Sum =",total) 
    
sum()
sum(10)
sum(10,20)
sum(10,20,30,40) 

The Sum = 0
The Sum = 10
The Sum = 30
The Sum = 100


___Note: We can mix variable length arguments with positional arguments.___

In [68]:
# Eg:
def f1(n1,*s):
    print(n1)
    for s1 in s:
        print(s1)

f1(10)
f1(10,20,30,40)
f1(10,"A",30,"B") 

10
10
20
30
40
10
A
30
B


___Note: After variable length argument,if we are taking any other arguments then we
should provide values as keyword arguments.___

In [69]:
# Eg:
def f1(*s,n1):
    for s1 in s:
        print(s1)
    print(n1)

f1("A","B",n1=10)

A
B
10


    f1("A","B",10) ==>Invalid
    TypeError: f1() missing 1 required keyword-only argument: 'n1'
___Note: We can declare key word variable length arguments also.
For this we have to use **.___
    
    def f1(**n):

We can call this function by passing any number of keyword arguments. Internally these
keyword arguments will be stored inside a dictionary.

In [70]:
# Eg:
def display(**kwargs):
    for k,v in kwargs.items():
        print(k,"=",v)
display(n1=10,n2=20,n3=30)
display(rno=100,name="Nikhil",marks=70,subject="Machine Learning") 

n1 = 10
n2 = 20
n3 = 30
rno = 100
name = Nikhil
marks = 70
subject = Machine Learning


### Case Study:
    def f(arg1,arg2,arg3=4,arg4=8):
        print(arg1,arg2,arg3,arg4)

    1. f(3,2) ==> 3 2 4 8
    2. f(10,20,30,40) ===>10 20 30 40
    3. f(25,50,arg4=100) ==>25 50 4 100
    4. f(arg4=2,arg1=3,arg2=4)===>3 4 4 2
    5. f()===>Invalid
TypeError: f() missing 2 required positional arguments: 'arg1' and 'arg2'

    6. f(arg3=10,arg4=20,30,40) ==>Invalid
 SyntaxError: positional argument follows keyword argument
 
 [After keyword arguments we should not take positional arguments]
    
    7. f(4,5,arg2=6)==>Invalid
 TypeError: f() got multiple values for argument 'arg2'
    
    8. f(4,5,arg3=5,arg5=6)==>Invalid
 TypeError: f() got an unexpected keyword argument 'arg5'

#### Note: Function vs Module vs Library:

1. A group of lines with some name is called a function
2. A group of functions saved to a file , is called Module
3. A group of Modules is nothing but Library

![](images\lib_fun.jpg)

### Types of Variables
Python supports 2 types of variables.

1. Global Variables
2. Local Variables

#### 1. Global Variables

The variables which are declared outside of function are called global variables.

These variables can be accessed in all functions of that module.

In [71]:
# Eg:
a=10 # global variable
def f1():
    print(a)

def f2():
    print(a)

f1()
f2() 

10
10


#### 2. Local Variables:
The variables which are declared inside a function are called local variables.

Local variables are available only for the function in which we declared it.i.e from outside
of function we cannot access.


#### global keyword:
We can use global keyword for the following 2 purposes:

1. To declare global variable inside function
2. To make global variable available to the function so that we can perform required modifications


In [72]:
# Eg 1:
a=10
def f1():
    a=777
    print(a)
def f2():
    print(a)
f1()
f2() 

777
10


In [73]:
# Eg 2:
a=10
def f1():
    global a
    a=555
    print(a)
def f2():
    print(a)
f1()
f2()

555
555


___Note: If global variable and local variable having the same name then we can access
global variable inside a function as follows___

In [74]:
a=10 #global variable
def f1():
    a=777 #local variable
    print(a)
    print(globals()['a'])
f1()

777
10


### Recursive Functions
A function that calls itself is known as Recursive Function.


#### Q. Write a Python Function to find factorial of given number with recursion.

In [75]:
def factorial(n):
    if n==0:
        result=1
    else:
        result=n*factorial(n-1)
    return result
print("Factorial of 4 is :",factorial(3))
print("Factorial of 5 is :",factorial(5)) 

Factorial of 4 is : 6
Factorial of 5 is : 120


### Anonymous Functions:
Sometimes we can declare a function without using dif or any name,such type of nameless functions are called anonymous functions or lambda functions.

The main purpose of anonymous function is just for instant use(i.e for one time usage)

### Normal Function:
We can define by using def keyword.

    def squareIt(n):
     return n*n
     
### lambda Function:
We can define by using lambda keyword lambda n:n*n

### Syntax of lambda Function:

    lambda parameter_list : expression
___Note: By using Lambda Functions we can write very concise code so that readability of the program will be improved.___

Lambda function return a function and hence they should be assigned to a function as: 
    
    f = lambda x : x*x

Here, 'f' is the function name to which the lambda expression is assigned. Now, if we call the function f() as:

    value = f(5)
    
Now, 'value' contains the square value of 5, i.e. 25. 

In [76]:
f = lambda x : x*x

In [77]:
f(5)

25

#### Q. Write a program to create a lambda function to find square of given number?

In [78]:
s = lambda n:n*n
print("The Square of 4 is :",s(5))
print("The Square of 5 is :",s(8)) 

The Square of 4 is : 25
The Square of 5 is : 64


#### Q. Lambda function to find sum of 2 given numbers

In [79]:
s=lambda a,b:a+b
print("The Sum of 10,20 is:",s(30,20))
print("The Sum of 50,200 is:",s(50,200))

The Sum of 10,20 is: 50
The Sum of 50,200 is: 250


#### Q. Lambda Function to find biggest of given values.

In [80]:
s=lambda a,b:a if a>b else b
print("The Biggest of 10,40 is:",s(10,40))
print("The Biggest of 300,200 is:",s(200,300)) 

The Biggest of 10,40 is: 40
The Biggest of 300,200 is: 300


In [81]:
grade = lambda x: 'A+' if x>7 else 'A' if x>=5 and x<=7 else 'B'

In [82]:
grade(8)

'A+'

In [83]:
lambda num: num**2

<function __main__.<lambda>(num)>

#### Note:
Lambda Function internally returns expression value and we are not required to write
return statement explicitly.
#### Note: 
Sometimes we can pass function as argument to another function. In such cases
lambda functions are best choice.

We can use lambda functions very commonly with filter(),map() and reduce() functions,b'z
these functions expect function as argument.

### filter() function:
The filter function is useful to filter out the elements os a sequence depending on the result of a function.

We can use filter() function to filter values from the given sequence based on some
condition.
    
    filter(function,sequence)
Where function argument is responsible to perform conditional check sequence can be list or tuple or string or set.


In [84]:
# Ex
letters = ['a', 'b', 'd', 'e', 'i', 'j', 'o']

# a function that returns True if letter is vowel
def filter_vowels(letter):
    vowels = ['a', 'e', 'i', 'o', 'u']
    return True if letter in vowels else False

filtered_vowels = filter(filter_vowels, letters)

# converting to tuple 'mandatory'
vowels = tuple(filtered_vowels)
print(vowels)

('a', 'e', 'i', 'o')


In [85]:
def is_even(x):
    if x%2==0:
        return "Correct"
    # We can also apply multiple condition to filter out the elements
    else:
        return False

In [91]:
is_even(3)

False

In [92]:
numbers = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16]
even_no = tuple(filter(is_even,numbers))
print(even_no)

(2, 4, 6, 8, 10, 12, 14, 16)


### map() function:
The map() function is similar to filter() function but it acts on each element of the sequence and perhaps changes the elements.

For every element present in the given sequence,apply some functionality and generate
new element with the required modification. For this requirement we should go for
map() function.

___Eg: For every element present in the list perform double and generate new list of doubles.___
#### Syntax:
    
    map(function,sequence)
The function can be applied on each element of sequence and generates new sequence.

In [94]:
numbers = [1, 2, 3, 4, 5, 6, 7]

# the lambda function returns True for even numbers 
f = lambda x: (x%2 == 0)
even_numbers_iterator = filter(f, numbers)

# converting to list
even_numbers = tuple(even_numbers_iterator)

print(even_numbers)

(2, 4, 6)


#### Without lambda

In [96]:
# Ex-2
listt=[1,2,3,4,5]
def doubleIt(x):
    return 2*x
l1=tuple(map(doubleIt,listt))
print(l1) 

(2, 4, 6, 8, 10)


#### with lambda

In [97]:
l=[1,2,3,4,5]
f = lambda x:2*x
l1=tuple(map(f,l))
print(l1) 

(2, 4, 6, 8, 10)


#### Eg 2: To find square of given numbers

In [98]:
l=[1,2,3,4,5]
l1=tuple(map(lambda x:x*x,l))
print(l1)

(1, 4, 9, 16, 25)


#### We can apply map() function on multiple lists also.But make sure all list should have same length. if the length ar'ent same than it work on the less list sequence of elements.

    Syntax: map(lambda x,y:x*y,l1,l2))
     x is from l1 and y is from l2

In [99]:
# Eg:
l1=[1,2,3,4]
l2=[2,3,4,5,]
l3=tuple(map(lambda x,y:x*y,l1,l2))
print(l3)

(2, 6, 12, 20)


### reduce() function:
reduce() function reduces sequence of elements into a single element by applying the specified function.

Here, the reduce function reduces the list to a final value as indicated by the lambda function.


    reduce(function,sequence)

___reduce() function present in functools module and hence we should write import statement.___


In [100]:
# import module and reduce fun
from functools import *

# Eg
l=[10,20,30,40,50]
result=reduce(lambda x,y:x+y,l)
print(result)

150


In [101]:
# Eg:
result=reduce(lambda x,y:x*y,l)
print(result) 

# Eg:
from functools import *
result=reduce(lambda x,y:x+y,range(1,101))
print(result)  

12000000
5050


#### Note:
- In Python every thing is treated as object.
- Even functions also internally treated as objects only.

In [102]:
# Eg:
def f1():
    print("Hello")
print(f1)
print(id(f1)) 

<function f1 at 0x0000021DC58A35E0>
2326891476448


### Function Aliasing:
For the existing function we can give another name, which is nothing but function aliasing. 

In [103]:
# Eg:
def wish(name):
    print("Good Morning:",name)
greeting=wish
print(id(wish))
print(id(greeting))
greeting('Nikhil')
wish('Prakriti') 

2326891475872
2326891475872
Good Morning: Nikhil
Good Morning: Prakriti


#### Note: 
In the above example only one function is available but we can call that function by using
either wish name or greeting name.

If we delete one name still we can access that function by using alias name

In [104]:
# Eg:
def wish(name):
    print("Good Morning:",name) 
greeting=wish
greeting('Nikhil')
wish('Prakriti')
del wish
            #wish('Nikhil') ==>NameError: name 'wish' is not defined
greeting('Yadav Ji')

Good Morning: Nikhil
Good Morning: Prakriti
Good Morning: Yadav Ji


### Nested Functions:
We can declare a function inside another function, such type of functions are called Nested functions.


In [105]:
# Eg:
def outer():
    print("outer function started")
    def inner():
        print("inner function execution")
    print("outer function calling inner function")
    inner()
outer()
#inner() ==>NameError: name 'inner' is not defined 

outer function started
outer function calling inner function
inner function execution


In the above example inner() function is local to outer() function and hence it is not possible to call
directly from outside of outer() function.

___Note: A function can return another function.___

In [106]:
# Eg:
def outer():
    print("outer function started")
    def inner():
        print("inner function execution") 
    print("outer function returning inner function")
    return inner
f1=outer()
f1()
f1()
f1() 

outer function started
outer function returning inner function
inner function execution
inner function execution
inner function execution


#### Q. What is the differenece between the following lines?
 
    f1 = outer
    f1 = outer()
    
- In the first case for the outer() function we are providing another name f1(function aliasing).
- But in the second case we calling outer() function,which returns inner function.For that inner function() we are providing another name f1

___Note: We can pass function as argument to another function___

    Eg: filter(function,sequence)
        map(function,sequence)
        reduce(function,sequence)

### Function Decorators:
Decorator is a function which can take a function as argument and extend its functionality
and returns modified function with extended functionality.

![](images\fun.jpg)

The main objective of decorator functions is we can extend the functionality of existing functions without modifies that function.

- We should define a decorator function with ___another function name as parameter___ as an example, let's define a decorator function decor() with 'fun'  as parameter.

        def decor(fun):

In [107]:
def wish(name):
    print("Hello",name,"Good Morning") 
    
# This function can always print same output for any name

But we want to modify this function to provide different message if name is Sunny.
- We can do this without touching wish() function by using decorator.
- So, the '@' symbol is useful to call the associated decorator internally.


In [108]:
# Eg: If we used different parameter name of a main function than we should use @decor for associated

def decor(func):    # (wish == func) - func is a main function is wish(), which is passing as parameter is the decor function.
    def inner(name):
        if name=="Sunny":
            print("Hello Sunny Bad Morning")
        else:
            func(name)    # arguments
    return inner
@decor
def wish(name): # parameter
    print("Hello",name,"Good Morning")
wish("Nikhil")
wish("Prakriti")
wish("Sunny") 

Hello Nikhil Good Morning
Hello Prakriti Good Morning
Hello Sunny Bad Morning


In [109]:
# Eg: If we used parameter name same as a main function name than we should not use @decor for associated

def decor(wish):  
    def inner(name):
        if name=="Sunny":
            print("Hello Sunny Bad Morning")
        else:
            wish(name)    # arguments
    return inner

def wish(name): # parameter
    print("Hello",name,"Good Morning")
wish("Nikhil")
wish("Prakriti")
wish("Sunny") 

Hello Nikhil Good Morning
Hello Prakriti Good Morning
Hello Sunny Good Morning


- In the above program whenever we call wish() function automatically decor function will
be executed.

#### How to call same function with decorator and without decorator:

#### We should not use @decor


In [110]:
def decor(func):     # func is a main function is wish(), which is passing as parameter is the decor fun
    def inner(name):
        if name=="Sunny":
            print("Hello Sunny Bad Morning")
        else:
            func(name) 
    return inner

def wish(jh):
    print("Hello",name,"Good Morning")
    
decorfunction=decor(wish)

wish("Nikhil") #decorator wont be executed
wish("Prakriti") #decorator wont be executed
decorfunction("Nikhil") #decorator will be executed
decorfunction("Sunny") #decorator will be executed 

Hello Prakriti Good Morning
Hello Prakriti Good Morning
Hello Prakriti Good Morning
Hello Sunny Bad Morning


In [111]:
# Eg 2:
def smart_division(func):
    def inner(a,b):
        print("We are dividing",a,"with",b)
        if b==0:
            print("OOPS...You cannot divide")
        else:
            return func(a,b) 
    return inner

@smart_division
def division(a,b):
    return a/b
print(division(20,2))
print(division(20,0)) 

We are dividing 20 with 2
10.0
We are dividing 20 with 0
OOPS...You cannot divide
None


#### with decorator we won't get any error. In this case output is:

    We are dividing 20 with 2
    10.0
    We are dividing 20 with 0
    OOPS...cannot divide
    None


### Decorator Chaining
We can define multiple decorators for the same function and all these decorators will
form Decorator Chaining.

In [112]:
# Eg:
def decor1(func):
    def inner():
        x=func()
        return x*x  
    return inner
def decor(func):
    def inner():
        x=func()
        return 2*x
    return inner

@decor1
@decor
def num():
    return 10
print(num()) 

400


#### Demo Program for decorator Chaining:

In [113]:
def decor(func):
    def inner(name):
        print("First Decor(decor) Function Execution")
        func(name)
    return inner

def decor1(func):
    def inner(name):
        print("Second Decor(decor1) Execution")
        func(name)
    return inner
@decor1
@decor
def wish(name):
    print("Hello",name,"Good Morning")
wish("Nikhil") 

Second Decor(decor1) Execution
First Decor(decor) Function Execution
Hello Nikhil Good Morning


### Generators
Generator is a function which is responsible to generate a sequence of values.

We can write generator functions just like ordinary functions, but it uses yield keyword to
return values.

In [114]:
# Eg 1:
def mygen():
    yield 'A'
    yield 'B'
    yield 'C'

g=mygen()
print(type(g))

print(next(g))
print(next(g))
print(next(g))
print(next(g)) 

<class 'generator'>
A
B
C


StopIteration: 

In [115]:
# Eg 2: To generate first n numbers:
def firstn(num):
    n=1
    while n<=num:
        yield n
        n=n+1
values=firstn(5)
print(next(values))
print(next(values))
print(next(values))

1
2
3


In [117]:
# We can convert generator into list as follows:
values=firstn(10)
l1=tuple(values)
print(l1)

(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)


### Advantages of Generator Functions:

1. when compared with class level iterators, generators are very easy to use
2. Improves memory utilization and performance.
3. Generators are best suitable for reading data from large number of large files
4. Generators work great for web scraping and crawling.


### Generators vs Normal Collections wrt Utilization:

#### Normal Collection:
    l=[x*x for x in range(10000000000000000)]
    print(l[0])
We will get MemoryError in this case because all these values are required to store in the memory.
#### Generators:
    g=(x*x for x in range(10000000000000000))
    print(next(g))
Output: 0

We won't get any MemoryError because the values won't be stored at the beginning

## Modules

### Math Module

Python has a built-in math module that can handle these complex calculations. Both simple mathematical calculations like addition (+), and subtraction (-), and advanced mathematical calculations like trigonometric operations, and logarithmic operations can be performed by the functions in the math module.

In [118]:
# This program will show the calculation of square root using the math module  
# importing the math module  
import math  
print(math.sqrt( 9 ))  

3.0


### Constants in Math Module
    
    pi    - The mathematical constant n=3.141592 with high precision.
    e     - The mathematical constant e=2.718281 with high precision.
    tau   - Ratio of the circumference of a circle to the radius of a circle.
    inf   - A floating-point positive infinity. (For negative infinity, use -math.inf).
    nan   - A floating-point "not a number" (NaN) value. Equivalent to the output of float('nam').

In [119]:
print(math.e)
print(math.pi)
print(math.tau)
print(math.inf)
print(math.nan)

2.718281828459045
3.141592653589793
6.283185307179586
inf
nan


### Description of all the Functions in Python Math Module
Here is a list of all the properties and functions specified in the math module, along with a brief description of what each one does.

    Function    ---->    Description
    
    ceil(x)    ---->    The lowest integer bigger than or equal to x is returned.
    copysign(x, y)    ---->    gives x back with the sign of y.
    fabs(x)    ---->    gives x's absolute value back.
    factorial(x)    ---->    provides the x factorial back.
    floor(x)    ---->    gives back the biggest integer that is less than or equal to x.
    fmod(x, y)    ---->    returns the leftover value after dividing x by y.
    frexp(x)    ---->    returns the pair of the mantissa and exponent of x. (m, e)
    fsum(iterable)    ---->    returns the iterable's correct floating point sum of all values.
    isfinite(x)    ---->    If x is neither an infinity nor a NaN, it returns True (Not a Number)
    isinf(x)    ---->    If x is a positive or negative infinity, it returns True.
    isnan(x)    ---->    If x is a NaN, it returns True.
    ldexp(x, i)    ---->    gives back x * (2**i).
    modf(x)    ---->    gives x's fractional and integer components back.
    trunc(x)    ---->    x's shortened integer value is returned.
    exp(x)    ---->    delivers e**x
    expm1(x)    ---->    yields e**x - 1
    log(x[, b])    ---->    gives back the x logarithm in base b. (defaults to e)
    log1p(x)    ---->    the natural logarithm of 1 + x is returned.
    log2(x)    ---->    gives x's base-2 logarithm back.
    log10(x)    ---->    provides x's base-10 logarithm.
    pow(x, y)    ---->    gives x raised to the power of y back.
    sqrt(x)    ---->    gives x's square root back.
    acos(x)    ---->    gives the arc cosine of x back.
    asin(x)    ---->    gives the arc sine of x back.
    atan(x)    ---->    gives the arc tangent of x back.
    atan2(y, x)    ---->    gives back atan(y / x).
    cos(x)    ---->    returns the x's cosine.
    hypot(x, y)    ---->    returns sqrt(x*x + y*y), the Euclidean norm.
    sin(x)    ---->    gives the sine of x back.
    tan(x)    ---->    gives the tangent of x back.
    degrees(x)    ---->    Angle x is transformed from radians to degrees.
    radians(x)    ---->    Angle x is transformed from degrees to radians.
    acosh(x)    ---->    x's inverse hyperbolic cosine is returned.
    asinh(x)    ---->    x's inverse hyperbolic sine is returned.
    atanh(x)    ---->    x's inverse hyperbolic tangent is returned.
    cosh(x)    ---->    gives x's hyperbolic cosine.
    sinh(x)    ---->    gives x's hyperbolic cosine.
    tanh(x)    ---->    gives x's hyperbolic tangent back.
    erf(x)    ---->    the error function at x is returned.
    erfc(x)    ---->    a function that gives the complementary error at x
    gamma(x)    ---->    the Gamma function at x is returned.
    lgamma(x)    ---->    gives the natural logarithm of the gamma function's absolute value at x.
    pi    ---->    The ratio of a circle's circumference to its diameter is a mathematical constant (3.14159...)
    e    ---->    e is a constant in mathematics (2.71828...)

A group of functions, variables and classes saved to a file, which is nothing but module.
 Every Python file (.py) acts as a module.

In [120]:
# Ex- module.py
x=800200
def add(a,b):
    print("The Sum:",a+b)
def product(a,b):
    print("The Product:",a*b)

In [121]:
import module as nikmod
nikmod.add(10,5)

The Sum: 15


This module contains one variable and 2 functions.
If we want to use members of module in our program then we should import that module.
#### import modulename
We can access members by using module name.

    modulename.variable
    modulename.function()

In [122]:
# test.py
import module

module.add(10,20)
module.product(10,20)

The Sum: 30
The Product: 200


#### Note:
whenever we are using a module in our program, for that module compiled file will be
generated and stored in the hard disk permanently.
Renaming a module at the time of import (module aliasing):
#### Eg:

    import module as m
- Here, module is original module name and m is alias name.
- We can access members by using alias name m

#### from ... import:
We can import particular members of module by using from ... import .

The main advantage of this is we can access members directly without using module
name.
#### Eg:
    from module import x,add
    print(x)
    add(10,20)
    product(10,20)==> NameError: name 'product' is not defined
We can import all members of a module as follows
from nikhilmath import *

### Various possibilties of import:

    import modulename
    import module1,module2,module3
    import module1 as m
    import module1 as m1,module2 as m2,module3
    from module import member
    from module import member1,member2,memebr3
    from module import memeber1 as x
    from module import *

#### Finding members of module by using dir() function:
Python provides inbuilt function dir() to list out all members of current module or a specified module.

    dir() ===>To list out all members of current module
    dir(moduleName)==>To list out all members of specified module

### The Special variable __name__:
For every Python program , a special variable __name__ will be added internally.
This variable stores information regarding whether the program is executed as an individual program or as a module.

If the program executed as an individual program then the value of this variable is
__main__

If the program executed as a module from some other program then the value of this variable is the name of module where it is defined.

Hence by using this __name__ variable we can identify whether the program executed directly or as a module.

#### Demo program:

In [123]:
# module1.py:
def f1():
    if __name__=='__main__':
        print("The code executed as a program")
    else:
        print("The code executed as a module from some other program")
f1() 

The code executed as a program


### Note:
We can find help for any module by using help() function
    
    Eg:
    import math
    help(math)

### Working with random module:
This module defines several functions to generate random numbers.
We can use these functions while developing games,in cryptography and to generate random numbers on fly for authentication.

### 1. random() function:
 This function always generate some float value between 0 and 1 ( not inclusive)
 
     0<x<1

In [124]:
# Eg:
from random import *
for i in range(5):
    print(random()) 

0.6837957913051469
0.832112246743524
0.30097866221791647
0.840248970525521
0.3265443692975978


### 2. randint() function:
To generate random integer beween two given numbers(inclusive)


In [125]:
# Eg:
from random import *
for i in range(10):
    print(randint(1,100)) # generate random int value between 1 and 100(inclusive) 

4
100
1
68
17
84
12
89
82
55


### 3. uniform():
It returns random float values between 2 given numbers(not inclusive)

In [126]:
# Eg:
from random import *
for i in range(5):
    print(uniform(1,10)) 

6.516221300776864
2.4601237902281854
7.788057082873847
6.630052157510197
2.2792601835163637


    random() ===>in between 0 and 1 (not inclusive)
    randint(x,y) ==>in between x and y ( inclusive)
    uniform(x,y) ==> in between x and y ( not inclusive)
    
### 4. randrange([start],stop,[step])

    returns a random number from range
        start<= x < stop
         start argument is optional and default value is 0
         step argument is optional and default value is 1
         
    randrange(10)-->generates a number from 0 to 9
    randrange(1,11)-->generates a number from 1 to 10
    randrange(1,11,2)-->generates a number from 1,3,5,7,9

In [127]:
# Eg 1:
from random import *
for i in range(5):
    print(randrange(10)) 

5
6
0
6
6


In [128]:
# Eg 2:
from random import *
for i in range(5):
    print(randrange(1,11))

9
9
10
8
8


In [129]:
# Eg 3:
from random import *
for i in range(5):
    print(randrange(1,11,2)) 

3
3
5
9
1


### 5. choice() function:
- It wont return random number.
- It will return a random object from the given list or tuple.

In [130]:
# Eg:
from random import *
list=["Nikhil","Akhil","Prakriti","Happy","Akhil"]
for i in range(8):
    print(choice(list)) 

Happy
Happy
Nikhil
Akhil
Akhil
Happy
Prakriti
Akhil
